File: java/ThingThroughput/src/writer/ThroughputWriter.java
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This is a simple throughput application measuring obtainable throughput using the thingapi
 *
 */

package writer;

import java.io.Closeable;
import java.util.concurrent.atomic.AtomicBoolean;

import com.adlinktech.datariver.DataRiver;
import com.adlinktech.datariver.DiscoveredThingRegistry;
import com.adlinktech.datariver.IotByteSeq;
import com.adlinktech.datariver.IotNvp;
import com.adlinktech.datariver.IotNvpSeq;
import com.adlinktech.datariver.IotValue;
import com.adlinktech.datariver.JSonTagGroupRegistry;
import com.adlinktech.datariver.JSonThingClassRegistry;
import com.adlinktech.datariver.JSonThingProperties;
import com.adlinktech.datariver.OutputHandler;
import com.adlinktech.datariver.Thing;
import com.adlinktech.datariver.ThingAPIRuntimeError;


public class ThroughputWriter implements Closeable {
	
	private AtomicBoolean stop = new AtomicBoolean(false);
	
	private final String thingPropertiesUri;
    private final DataRiver dataRiver;
    private final Thing thing;
    private final IotNvpSeq sample;
	
	public ThroughputWriter(final String thingPropertiesURI) {
		System.out.println("Throughput writer started");
		this.thingPropertiesUri = thingPropertiesURI;
		this.dataRiver = DataRiver.getInstance();
		this.thing = createThing();
		this.sample = new IotNvpSeq();
	}
	
	public synchronized void registerCtrlHandler() {
        final Thread mainThread = Thread.currentThread();
        Runtime.getRuntime().addShutdownHook(new Thread() {
            public void run() {
                stop.set(true);
                while (true) {
                    try {
                        mainThread.join();
                        return;
                    } catch (InterruptedException e) {
                        // someone tried to stop us, still wait for mainThread
                    }
                }
            }
        });
    }
	
	private void cleanup() {
		if(this.dataRiver != null) {
			this.dataRiver.close();
		}
		System.out.println("Throughput writer stopped");
	}
	
	private String getAbsFileURI(final String fileName) {
	    final String dir = System.getProperty("user.dir");
	    final String prefix = "file://";
		return prefix.concat(dir).concat("/").concat(fileName);
	}
	
	private Thing createThing() {
	    final JSonTagGroupRegistry tgr = new JSonTagGroupRegistry();
		tgr.registerTagGroupsFromUri(getAbsFileURI("definitions/TagGroup/com.adlinktech.example/ThroughputTagGroup.json"));
		this.dataRiver.addTagGroupRegistry(tgr);
		
		final JSonThingClassRegistry tcr = new JSonThingClassRegistry();
		tcr.registerThingClassesFromUri(getAbsFileURI("definitions/ThingClass/com.adlinktech.example/ThroughputWriterThingClass.json"));
		this.dataRiver.addThingClassRegistry(tcr);
		
		final JSonThingProperties tp = new JSonThingProperties();
		tp.readPropertiesFromUri(this.thingPropertiesUri);
		
		return this.dataRiver.createThing(tp);
	}

    private void setupMessage(final long payloadSize) {
        final IotValue sequenceDataValue = new IotValue();
        final IotValue sequenceNumberValue = new IotValue();
    	
    	// Init data Tag
        final IotByteSeq sequenceData = new IotByteSeq();
    	for(long i = 0; i < payloadSize; i++) {
    		sequenceData.add((byte) Character.getNumericValue('a'));
    	}
    	sequenceDataValue.setByteSeq(sequenceData);
    	
    	
    	// Init sequence number and size Tags
    	sequenceNumberValue.setUint64((long) 0);

        // Create sample
    	this.sample.add(new IotNvp("sequencenumber", sequenceNumberValue));
    	this.sample.add(new IotNvp("sequencedata", sequenceDataValue));
    }

    private void waitForReader() throws InterruptedException {
        // wait for throughputreader to appear by discovering its thingId and thingClass
    	System.out.println("Waiting for Throughput reader.. ");
    	final DiscoveredThingRegistry discoveredThingRegistry = this.dataRiver.getDiscoveredThingRegistry();
    	
        while (!stop.get()) {
            try {
                discoveredThingRegistry.findDiscoveredThing("*", "ThroughputReader:com.adlinktech.example:v1.0");
                break;
            } catch (ThingAPIRuntimeError e) {
                // Thing not available yet. Sleep for 100 milliseconds and try again.
            }
            
            Thread.sleep(100);
        }

        if (!stop.get()) {
        	System.out.println("Throughput reader found");
        } else {
        	System.out.println("Terminated");
        	System.exit(1);
        }
    }


    private void write(
            final long burstInterval,
            final long burstSize,
            final long runningTime,
            final WriterMode mode) throws InterruptedException {
        
        long burstCount = 0;
        long count = 0;
        boolean timedOut = false;

        final long pubStart = System.currentTimeMillis();
        long burstStart = System.currentTimeMillis();
        long currentTime = System.currentTimeMillis();

        final OutputHandler outputHandler = this.thing.getOutputHandler("ThroughputOutput");
        IotValue internalSequenceNumberValue = new IotValue();

        if (mode == WriterMode.OUTPUT_HANDLER_NOT_THREAD_SAFE) {
            outputHandler.setNonReentrantFlowId(this.thing.getContextId());
            final IotNvpSeq internalNvpSeq = outputHandler.setupNonReentrantNvpSeq(this.sample);

            final IotNvp internalSequenceNumberNvp = internalNvpSeq.get(0);
            internalSequenceNumberValue = internalSequenceNumberNvp.getValue();
        }

        while (!stop.get() && !timedOut)
        {
            // Write data until burst size has been reached
            if (burstCount++ < burstSize) {
                if (mode == WriterMode.OUTPUT_HANDLER) {
                    // Fill the nvp_seq with updated sequencenr
                    this.sample.get(0).getValue().setUint64(count++);

                    // Write the data using output handler
                    outputHandler.write(this.sample);
                } else if (mode == WriterMode.OUTPUT_HANDLER_NOT_THREAD_SAFE) {
                    // Fill the nvp_seq with updated sequencenr
                	internalSequenceNumberValue.setUint64(count++);

                    // Write the data using non-reentrant write on output handler
                    outputHandler.writeNonReentrant();
                } else {
                    // Fill the nvp_seq with updated sequencenr
                	this.sample.get(0).getValue().setUint64(count++);

                    // Write the data
                    this.thing.write("ThroughputOutput", this.sample);
                }
            } else if (burstInterval != 0) {
                // Sleep until burst interval has passed
                currentTime = System.currentTimeMillis();

                final long deltaTime = currentTime - burstStart;
                if (deltaTime < burstInterval) {
                	Thread.sleep(burstInterval - deltaTime);
                }
                burstStart = System.currentTimeMillis();
                burstCount = 0;
            } else {
                burstCount = 0;
            }

            // Check of timeout
            if (runningTime != 0) {
                currentTime = System.currentTimeMillis();
                if (((float)(currentTime - pubStart) / 1000.0) > runningTime) {
                    timedOut = true;
                }
            }
        }

        // Show stats
        if (stop.get()) {
        	System.out.println("Terminated: " + count + " samples written");
        } else {
        	System.out.println("Timed out: " + count + " samples written");
        }
    }
    
    public void run(final long payloadSize, final long burstInterval, final long burstSize, final long runningTime, final WriterMode writerMode) {
        final String writerModeStr = (writerMode == WriterMode.OUTPUT_HANDLER) ?
            "outputHandler" : ((writerMode == WriterMode.OUTPUT_HANDLER_NOT_THREAD_SAFE) ?
                "outputHandlerNotThreadSafe" : "standard");

        System.out.println("payloadSize: " + payloadSize + " | burstInterval: " + burstInterval
            + " | burstSize: " + burstSize + " | runningTime: " + runningTime
            + " | writer-mode: " + writerModeStr);
        
        try {
            // Wait for reader to be discovered
            waitForReader();
    
            // Create the message that is sent
            setupMessage(payloadSize);
    
            // Write data
            write(burstInterval, burstSize, runningTime, writerMode);
    
            // Give middleware some time to finish writing samples
            Thread.sleep(2000);
            
		} catch (InterruptedException e) {
		    System.out.println("ThroughputWriter intertrupted. Exiting");
		}
    }

    @Override
    public void close() {
        cleanup();
    }
    
}

File: dotnet/ThingThroughput/Writer/ThroughputWriter.cs
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
using System;
using Adlinktech.DataRiver;

namespace Writer
{
    public class ThroughputWriter : IDisposable
    {
        public bool Stop { get; set; }
        private DataRiver DataRiver { get; }
        private Thing Thing { get; }
        private IotNvpSeq Sample { get; }
        public WriterMode OutputHandlerNotThreadSafe { get; private set; }

        private void CtrlCHandler(object sender, ConsoleCancelEventArgs args)
        {
            Console.WriteLine("Writer trapped Ctrl-C. Stopping.");
            Stop = true;
            args.Cancel = true; // don't terminate the process, we'll do that.
        }

        private static Thing CreateThing(in DataRiver dataRiver)
        {
            var tgr = new JSonTagGroupRegistry();
            tgr.RegisterTagGroupsFromString(Utilities.GetEmbeddedResourceContent("Writer.definitions.TagGroup.com.adlinktech.example.ThroughputTagGroup.json"));
            dataRiver.AddTagGroupRegistry(tgr);

            var tcr = new JSonThingClassRegistry();
            tcr.RegisterThingClassesFromString(Utilities.GetEmbeddedResourceContent("Writer.definitions.ThingClass.com.adlinktech.example.ThroughputWriterThingClass.json"));
            dataRiver.AddThingClassRegistry(tcr);

            var tp = new JSonThingProperties();
            tp.ReadPropertiesFromString(Utilities.GetEmbeddedResourceContent("Writer.config.ThroughputWriterProperties.json"));

            return dataRiver.CreateThing(tp);
        }

        public ThroughputWriter()
        {
            DataRiver = DataRiver.GetInstance();
            Thing = CreateThing(DataRiver);
            Stop = false;
            Sample = new IotNvpSeq(2);

            Console.CancelKeyPress += this.CtrlCHandler;
            Console.WriteLine("Throughput writer started");
        }

        public int Run(in Options options)
        {
            Console.WriteLine($"payloadSize: {options.PayloadSize} " +
                $"| burstInterval: {options.BurstInterval} " +
                $"| burstSize: {options.BurstSize} " +
                $"| runningTime: {options.RunningTime} " +
                $"| writer-mode: {options.WriterMode}");

            // wait for a reader to be discovered
            WaitForReader();

            // create the message that is to be sent
            SetupMessage(options.PayloadSize);

            // write the data
            Write(options.BurstInterval, options.BurstSize, options.RunningTime, options.WriterMode);

            // sleep for two seconds to allow middleware to finish writing samples
            System.Threading.Thread.Sleep(2000);

            return 0;
        }

        private void WaitForReader()
        {
            Console.WriteLine("Waiting for Throughput reader.. ");
            var discoveredThingRegistry = DataRiver.DiscoveredThingRegistry;
            bool readerFound = false;
            while (!readerFound && !Stop)
            {
                try
                {
                    var thing = discoveredThingRegistry.FindDiscoveredThing("*", "ThroughputReader:com.adlinktech.example:v1.0");
                    readerFound = true;
                }
                catch(ThingAPIRuntimeError)
                {
                    // No result from FindDiscoveredThing
                    // Sleep 100ms and try again
                    System.Threading.Thread.Sleep(100);
                }
            }
            Console.WriteLine(Stop ? "Terminated" : "Throughput reader found");
        }

        private void SetupMessage(ulong payloadSize)
        {
            var sequenceData = new IotByteSeq((int)payloadSize);
            for (ulong i = 0; i < payloadSize; i++)
            {
                sequenceData.Add((byte)'a');
            }

            var snValue = new IotValue();
            snValue.UInt64 = 0;
            Sample.Add(new IotNvp("sequencenumber", snValue));
            var sdValue = new IotValue();
            sdValue.ByteSeq = sequenceData;
            Sample.Add(new IotNvp("sequencedata", sdValue));
        }

        private void Write(ulong burstInterval, ulong burstSize, ulong runningTime, WriterMode mode)
        {
            ulong burstCount = 0;
            uint count = 0;
            bool timedOut = false;

            var pubStart = DateTime.Now;
            var burstStart = DateTime.Now;

            OutputHandler outputHandler = Thing.GetOutputHandler("ThroughputOutput");
            IotValue internalSequencenumberV = null;

            if (mode == WriterMode.outputHandlerNotThreadSafe)
            {
                outputHandler.NonReentrantFlowId = Thing.ContextId;
                var internalNvpSeq = outputHandler.SetupNonReentrantNvpSeq(Sample);
                internalSequencenumberV = internalNvpSeq[0].Value;
            }

            while (!Stop && !timedOut)
            {
                // Write data until burst size has been reached
                if (burstCount++ < burstSize)
                {
                    if (mode == WriterMode.outputHandler)
                    {
                        // Fill the nvp_seq with updated sequencenr
                        Sample[0].Value.UInt64 = count++;

                        // Write the data using output handler
                        outputHandler.Write(Sample);
                    }
                    else if (mode == WriterMode.outputHandlerNotThreadSafe)
                    {
                        // Fill the nvp_seq with updated sequencenr
                        internalSequencenumberV.UInt64 = count++;

                        // Write the data using non-reentrant write on output handler
                        outputHandler.WriteNonReentrant();
                    }
                    else
                    {
                        // Fill the nvp_seq with updated sequencenr
                        Sample[0].Value.UInt64 = count++;

                        // Write the data
                        Thing.Write("ThroughputOutput", Sample);
                    }
                }
                else if (burstInterval != 0)
                {
                    // Sleep until burst interval has passed
                    TimeSpan deltaTime = DateTime.Now - burstStart;
                    var remainingBurstIntervalMillis = (int)burstInterval
                        - (int)deltaTime.TotalMilliseconds;
                    if (remainingBurstIntervalMillis > 0)
                    {
                        System.Threading.Thread.Sleep(remainingBurstIntervalMillis);
                    }
                    burstStart = DateTime.Now;
                    burstCount = 0;
                }
                else
                {
                    burstCount = 0;
                }

                // Check of timeout
                if (runningTime != 0)
                {
                    TimeSpan deltaTime = DateTime.Now - pubStart;
                    var remainingSeconds = (int)runningTime - deltaTime.TotalSeconds;
                    if (remainingSeconds <= 0)
                    {
                        timedOut = true;
                    }
                }
            }


            // Show stats
            if (Stop)
            {
                Console.WriteLine($"Terminated: {count} samples written");
            }
            else
            {
                Console.WriteLine($"Timed out: {count} samples written");
            }
        }

        public void Dispose()
        {
            Thing.Purge("ThroughputOutput");
            DataRiver.Close();
            DataRiver.Dispose();
            Console.WriteLine("Throughput writer stopped");
        }
    }
}
